src/backend.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (C) Nitrokey GmbH
+// SPDX-License-Identifier: Apache-2.0 or MIT
+
+mod data;
+
+use trussed::{
+    backend::Backend,
+    error::Result,
+    platform::Platform,
+    serde_extensions::ExtensionImpl,
+    service::ServiceResources,
+    store::filestore::Filestore,
+    types::{CoreContext, Location, PathBuf},
+};
+
+use crate::{
+    extension::{reply, AuthExtension, AuthReply, AuthRequest},
+    PIN_PATH,
+};
+use data::PinData;
+
+pub struct AuthBackend {
+    location: Location,
+}
+
+impl AuthBackend {
+    pub fn new(location: Location) -> Self {
+        Self { location }
+    }
+}
+
+impl<P: Platform> Backend<P> for AuthBackend {
+    type Context = ();
+}
+
+impl<P: Platform> ExtensionImpl<AuthExtension, P> for AuthBackend {
+    fn extension_request(
+        &mut self,
+        core_ctx: &mut CoreContext,
+        _ctx: &mut (),
+        request: &AuthRequest,
+        resources: &mut ServiceResources<P>,
+    ) -> Result<AuthReply> {
+        let fs = &mut resources.filestore(core_ctx);
+        match request {
+            AuthRequest::HasPin(request) => {
+                let has_pin = fs.exists(&request.id.path(), self.location);
+                Ok(reply::HasPin { has_pin }.into())
+            }
+            AuthRequest::CheckPin(request) => {
+                // TODO: return error or bool?
+                PinData::load(fs, self.location, request.id)?
+                    .write(fs, self.location, |data| data.check_pin(&request.pin))?;
+                Ok(reply::CheckPin.into())
+            }
+            AuthRequest::SetPin(request) => {
+                let mut rng = resources.rng().map_err(|_| Error::RngFailed)?;
+                PinData::new(request.id, &request.pin, request.retries, &mut rng)
+                    .save(fs, self.location)?;
+                Ok(reply::SetPin.into())
+            }
+            AuthRequest::DeletePin(request) => {
+                // TODO: what to do with missing files?
```

Comment:
it should deliver the promise that either there is no pin after this operation or it failed ensuring it, so:
* check for existence
* fail silently (result are equal: pin is deleted)